runners/map-reduce/src/main/java/org/apache/beam/runners/mapreduce/translation/BeamReducer.java

```diff
@@ -0,0 +1,104 @@
+import java.util.List;
+import org.apache.beam.sdk.coders.Coder;
+import org.apache.beam.sdk.coders.KvCoder;
+import org.apache.beam.sdk.util.SerializableUtils;
+import org.apache.beam.sdk.util.WindowedValue;
+import org.apache.beam.sdk.values.KV;
+import org.apache.commons.codec.binary.Base64;
+import org.apache.hadoop.io.BytesWritable;
+import org.apache.hadoop.mapreduce.Mapper;
+import org.apache.hadoop.mapreduce.Reducer;
+import org.apache.hadoop.mapreduce.TaskInputOutputContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Adapter for executing Beam transforms in {@link Reducer}.
+ */
+public class BeamReducer<ValueInT, ValueOutT>
+    extends Reducer<BytesWritable, byte[], Object, WindowedValue<ValueOutT>> {
+  private static final Logger LOG = LoggerFactory.getLogger(Reducer.class);
+
+  public static final String BEAM_REDUCER_KV_CODER = "beam-reducer-kv-coder";
+  public static final String BEAM_PAR_DO_OPERATION_REDUCER = "beam-par-do-op-reducer";
+
+  private Coder<Object> keyCoder;
+  private Coder<Object> valueCoder;
+  private Operation operation;
+
+  @Override
+  protected void setup(
+      Reducer<BytesWritable, byte[], Object, WindowedValue<ValueOutT>>.Context context) {
+    String serializedValueCoder = checkNotNull(
+        context.getConfiguration().get(BEAM_REDUCER_KV_CODER),
+        BEAM_REDUCER_KV_CODER);
+    String serializedParDo = checkNotNull(
+        context.getConfiguration().get(BEAM_PAR_DO_OPERATION_REDUCER),
+        BEAM_PAR_DO_OPERATION_REDUCER);
+    KvCoder<Object, Object> kvCoder = (KvCoder<Object, Object>) SerializableUtils
+        .deserializeFromByteArray(Base64.decodeBase64(serializedValueCoder), "Coder");
+    keyCoder = kvCoder.getKeyCoder();
+    valueCoder = kvCoder.getValueCoder();
+    operation = (Operation) SerializableUtils.deserializeFromByteArray(
+        Base64.decodeBase64(serializedParDo), "Operation");
+    operation.start((TaskInputOutputContext) context);
+  }
+
+  @Override
+  protected void reduce(
+      BytesWritable key,
+      Iterable<byte[]> values,
+      Reducer<BytesWritable, byte[], Object, WindowedValue<ValueOutT>>.Context context)
+      throws InterruptedException, IOException {
+    List<Object> decodedValues = Lists.newArrayList(FluentIterable.from(values)
+        .transform(new Function<byte[], Object>() {
+          @Override
+          public Object apply(byte[] input) {
+            ByteArrayInputStream inStream = new ByteArrayInputStream(input);
+            try {
+              return valueCoder.decode(inStream);
+            } catch (IOException e) {
+              Throwables.throwIfUnchecked(e);
+              throw new RuntimeException(e);
+            }
+          }}));
+    Object decodedKey = keyCoder.decode(new ByteArrayInputStream(key.getBytes()));
+    LOG.info("key: {} value: {}.", decodedKey, decodedValues);
+    operation.process(
+        WindowedValue.valueInGlobalWindow(KV.of(decodedKey, decodedValues)));
```

Comment:
<!--new_thread; commit:5079322c2e2a092a85b9740d04a7ca9bd887460e; resolved:0-->
Is this in the global window because it is always reading a globally windowed shuffle and will go through GABW? Or is there some limitation in the runner?